### PR TITLE
Provider: Signature verification

### PR DESCRIPTION
**Pull request by iaik-jheher**

### CODE REVIEW

File: provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/SignatureInput.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    constructor(data: ByteArray) : this(sequenceOf(data), RAW_BYTES)
+    constructor(data: Sequence<ByteArray>): this(data, RAW_BYTES)
+
+    /**
+     * Takes the leftmost bits of the byte array, and converts them to an unsigned `BigInteger`.
+     *
+     * (This matches the ECDSA spec.)
+     */
+    fun asBigInteger(length: BitLength): BigInteger {
```
Comment (JesusMcCloud) [inline provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/SignatureInput.kt:43]: KmmResult?

File: provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/Verifier.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    sealed class EC
+    @Throws(IllegalArgumentException::class)
+    constructor (
+        final override val signatureAlgorithm: SignatureAlgorithm.ECDSA,
+        final override val publicKey: CryptoPublicKey.EC)
+    : Verifier {
+        init {
+            signatureAlgorithm.requiredCurve?.let {
+                require(publicKey.curve == it)
```
Comment (JesusMcCloud) [inline provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/Verifier.kt:26]: Can we McGuyer a creation function that returns a KmmResult? not sure myself whether we want this

**Review by JesusMcCloud — COMMENTED** [review]
looking good so far

File: provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/Verifier.kt (hunk 2 above)
Comment (iaik-jheher) [reply provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/Verifier.kt:27]: i don't think this is a good idea in the constructor, but i also don't think the constructor will be the "canonical" way to construct verifiers. once we end up with `X509Algorithm.EC.verifierFor(publicKey)` then that can return `KmmResult`

File: provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/SignatureInput.kt (hunk 1 above)
Comment (iaik-jheher) [reply provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/SignatureInput.kt:43]: there's no useful error conditions to catch

Comment (JesusMcCloud) [reply provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/SignatureInput.kt:43]: Soooo it's an IO error?!

File: provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/Verifier.kt (hunk 2 above)
Comment (JesusMcCloud) [reply provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/Verifier.kt:27]: Then be sure to add API Docs, referencing `X509Algorithm.EC.verifierFor(publicKey)` once its there.

Comment (iaik-jheher) [reply provider/src/commonMain/kotlin/at/asitplus/crypto/provider/sign/Verifier.kt:27]: Sure, the constructor will likely end up internal anyway.

**Pull request merged**